File: android/src/main/kotlin/com/effdel/native_perms/NativePermsPlugin.kt
package com.effdel.native_perms

import android.content.Intent
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry

class NativePermsPlugin :
    FlutterPlugin,
    ActivityAware,
    MethodChannel.MethodCallHandler,
    PluginRegistry.RequestPermissionsResultListener,
    PluginRegistry.ActivityResultListener {

    private lateinit var channel: MethodChannel
    private lateinit var manager: PermissionManager
    private var activityBinding: ActivityPluginBinding? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "dev.effdel.native_perms/methods")
        channel.setMethodCallHandler(this)
        manager = PermissionManager(binding.applicationContext)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        manager.setActivity(binding.activity)
        binding.addRequestPermissionsResultListener(this)
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        detachActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        detachActivity()
    }

    private fun detachActivity() {
        activityBinding?.removeRequestPermissionsResultListener(this)
        activityBinding?.removeActivityResultListener(this)
        activityBinding = null
        manager.setActivity(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "checkPermissionStatus" -> {
                val perm = call.arguments as Int
                result.success(manager.checkPermissionStatus(perm))
            }
            "requestPermissions" -> {
                @Suppress("UNCHECKED_CAST")
                val perms = (call.arguments as List<Int>)
                manager.requestPermissions(perms) { map ->
                    result.success(map)
                }
            }
            "shouldShowRequestPermissionRationale" -> {
                val perm = call.arguments as Int
                result.success(manager.shouldShowRequestRationale(perm))
            }
            "checkServiceStatus" -> {
                val perm = call.arguments as Int
                val ctx = activityBinding?.activity ?: return result.success(
                    PermissionMapper.SERVICE_NOT_APPLICABLE,
                )
                result.success(ServiceManager.checkServiceStatus(ctx, perm))
            }
            "openAppSettings" -> {
                val ctx = activityBinding?.activity ?: return result.success(false)
                result.success(AppSettingsManager.openAppSettings(ctx))
            }
            else -> result.notImplemented()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray,
    ): Boolean = manager.onRequestPermissionsResult(requestCode, permissions, grantResults)

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean =
        manager.onActivityResult(requestCode, resultCode, data)
}

File: ios/nativeprems/Sources/nativeprems/NativePermsPlugin.swift
import Flutter
import UIKit

public class NativePermsPlugin: NSObject, FlutterPlugin {

    private let manager = PermissionManager()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "dev.effdel.nativeprems/methods",
            binaryMessenger: registrar.messenger()
        )
        let instance = NativePermsPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "checkPermissionStatus":
            guard let perm = call.arguments as? Int else {
                result(FlutterError.invalidArgs); return
            }
            manager.checkPermissionStatus(perm) { result($0) }

        case "requestPermissions":
            guard let perms = call.arguments as? [Int] else {
                result(FlutterError.invalidArgs); return
            }
            manager.requestPermissions(perms) { map in
                // Convert [Int: Int] to a dict the Flutter codec can carry.
                var out: [Int: Int] = [:]
                for (k, v) in map { out[k] = v }
                result(out)
            }

        case "shouldShowRequestPermissionRationale":
            // Android-only; Dart short-circuits but be defensive.
            result(false)

        case "checkServiceStatus":
            guard let perm = call.arguments as? Int else {
                result(FlutterError.invalidArgs); return
            }
            ServiceManager.checkServiceStatus(perm) { result($0) }

        case "openAppSettings":
            result(AppSettingsManager.openAppSettings())

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

private extension FlutterError {
    static var invalidArgs: FlutterError {
        FlutterError(code: "INVALID_ARGS", message: "Argument was not the expected type", details: nil)
    }
}
